### DISCUSSION TITLE
Version 2.1.1

### DISCUSSION THREAD

**Discussion by saran-t** [google-deepmind/mujoco#92] [Announcements]
## API changes

1. Added ``mj_printFormattedModel``, which accepts a format string for floating point numbers, for example to increase precision.
2. Added ``mj_versionString``, which returns human-readable string that represents the version of the MuJoCo binary.
3. Converted leading underscores to trailing underscores in private instances of API struct definitions, to conform to reserved identifier directive, see [C standard: Section 7.1.3](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf).

   This is a minor breaking change. Code which references private instances will break. To fix, replace leading underscores with trailing underscores, e.g. `struct _mjModel` becomes `struct mjModel_`.

## General

4. Safer string handling: replaced `strcat`, `strcpy`, and `sprintf` with `strncat`, `strncpy`, and
   `snprintf` respectively.
5. Changed indentation from 4 spaces to 2 spaces everywhere.

## Bug Fixes

6. Fixed reading from uninitialized memory in PGS solver.
7. Computed capsule inertias are now exact. Until this change, capsule masses and inertias computed by the
   [`compiler`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#compiler)'s `inertiafromgeom` mechanism were approximated by a cylinder, formed by the capsule's cylindrical middle section, extended on both ends by half the capsule radius. Capsule inertias are now computed with the [Parallel Axis theorem](https://en.wikipedia.org/wiki/Parallel_axis_theorem), applied to the two hemispherical end-caps.

   This is a minor breaking change. Simulation of a model with automatically-computed capsule inertias will be numerically different, leading to, for example, breakage of golden-value tests.
8. Fixed bug related to [`force`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-force) and [`torque`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-force) sensors. Until this change, forces torques reported by F/T sensors ignored out-of-tree constraint wrenches except those produced by contacts. Force and and torque sensors now correctly take into account the effects of [`connect`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#equality-connect) and [`weld`](https://mujoco.readthedocs.io/en/latest/XMLreference.html#equality-connect) constraints.

   Forces generated by [spatial tendons](https://mujoco.readthedocs.io/en/latest/XMLreference.html#sensor-force) which are outside the kinematic tree (i.e. between bodies which have no ancestral relationship) are still not taken into account by force and torque sensors. This remains a future work item.

## Code samples

9. `testspeed`: Added injection of pseudo-random control noise, turned on by default. This is to avoid settling into some fixed contact configuration and providing an unrealistic timing measure.
10. `simulate`:

    - Added slower-than-real-time functionality, which is controlled via the '+' and '-' keys.
    - Added sliders for injecting Brownian noise into the controls.
    - Added "Print Camera" button to print an MJCF clause with the pose of the current camera.
    - The camera pose is not reset when reloading the same model file.

## Updated dependencies

11. `TinyXML` was replaced with `TinyXML2` 6.2.0.
12. `qhull` was upgraded to version 8.0.2.
13. `libCCD` was upgraded to version 1.4.
14. On Linux, `libstdc++` was replaced with `libc++`.

## Binary build

15. MacOS packaging. We now ship Universal binaries that natively support both Apple Silicon and Intel CPUs.

    - MuJoCo library is now packaged as a [Framework Bundle](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPFrameworks/Concepts/FrameworkAnatomy.html), allowing it to be incorporated more easily into Xcode projects (including Swift projects). Developers are encouraged to compile and link against MuJoCo using the `-framework mujoco` flag, however all header files and the ``libmujoco.2.1.1.dylib` library can still be directly accessed inside the framework.
    - Sample applications are now packaged into an Application Bundle called `MuJoCo.app`. When launched via GUI, the bundle launches the `simulate` executable. Other precompiled sample programs are shipped inside that bundle (in `MuJoCo.app/Contents/MacOS`) and can be launched via command line.
    - Binaries are now signed and the disk image is notarized.

16. Windows binaries and libraries are now signed.
17. Link-time optimization is enabled on Linux and macOS, leading to an average of ~20% speedup when benchmarked on
   three test models (`cloth.xml`, `humanoid.xml`, and `humanoid100.xml`).
18. Linux binaries are now built with LLVM/Clang instead of GCC.
19. An AArch64 (aka ARM64) Linux build is also provided.
20. Private symbols are no longer stripped from shared libraries on Linux and MacOS.

## Sample models

21. Clean-up of the `model/` directory.

    - Rearranged into subdirectories which include all dependencies.
    - Added descriptions in XML comments, cleaned up XMLs.
    - Deleted some composite models: `grid1`, `grid1pin`, `grid2`, `softcylinder`, `softellipsoid`.

22. Added descriptive animations in `docs/images/models/`

<hr /><em>This discussion was created from the release <a href='https://github.com/deepmind/mujoco/releases/tag/2.1.1'>2.1.1</a>.</em>

**Comment by pab47** [comment]
It's not clear how to compile and execute the new releases. I am looking at 2.1.1 here: https://github.com/deepmind/mujoco/releases
The source files don't have libraries and there is no bin folder. 

I checked the "getting started" in the documentation here: https://mujoco.readthedocs.io/en/latest/programming.html?highlight=getting%20started#getting-started
It doesn't seem to explain how to use the latest releases (e.g., 2.1.1).

I am familiar with mujoco200 downloaded from roboti.us using the linux as well as the mac libraries, but can't seem to make the switch to 2.1.1

**Reply by saran-t** [reply]
The "source" download is something that GitHub automatically generates (we cannot disable this) and is just an archive of the GitHub repo. Since we haven't open sourced MuJoCo yet, the source package isn't particularly useful right now. You should download one of the platform-specific binary packages where you'll find precompiled libraries and sample programs.

**Comment by arschoon** [comment]
Are there any updates on the python binaries that were announced to be released?

**Reply by saran-t** [reply]
We have launched the new Python bindings as part of the [2.1.2 release](https://github.com/deepmind/mujoco/discussions/185) today. Let us know if you run into any issue!

**Reply by arschoon** [reply]
Woohoo, thank you!

**Comment by AIBoticsGmbH** [comment]
Probably not worth mentioning but the Parallel axis theorem link is broken.
https://en.wikipedia.org/wiki/Parallel_axis_theorem%3E vs https://en.wikipedia.org/wiki/Parallel_axis_theorem

Just thought I'd let you know

**Reply by saran-t** [reply]
Can you send us a PR to fix this? :)

**Reply by AIBoticsGmbH** [reply]
I'm referring to **this release notes**. I'm not sure I can send a pull request for that. Or can I ?

**Reply by saran-t** [reply]
Ah, sorry, I thought you were talking about the changelog in our documentation. This is now fixed, thanks!
